src/librustc_typeck/collect.rs

```diff
@@ -2735,3 +2737,72 @@ fn check_target_feature_trait_unsafe(tcx: TyCtxt<'_>, id: LocalDefId, attr_span:
         }
     }
 }
+
+fn traits_implemented_by<'a>(tcx: TyCtxt<'a>, type_: DefId) -> &'a [DefId] {
+    // There isn't a cheap way to do this. Just look at every trait!
+    let all_traits = tcx.all_traits(LOCAL_CRATE).iter().copied();
+    traits_implemented_impl(tcx, type_, all_traits)
+}
+
+fn traits_in_scope_implemented_by<'a>(
+    tcx: TyCtxt<'a>,
+    (module, type_): (LocalDefId, DefId),
+) -> &'a [DefId] {
+    let module = tcx.hir().as_local_hir_id(module);
+    let traits_in_scope: FxHashSet<_> = match tcx.in_scope_traits(module) {
+        Some(x) => x.iter().map(|candidate| candidate.def_id).collect(),
+        None => {
+            debug!("in_scope_traits returned None");
+            // TODO: maybe this should return no traits instead?
+            return traits_implemented_by(tcx, type_);
+        }
+    };
+    debug!("the following traits were in scope: {:?} (in module {:?})", traits_in_scope, module);
+
+    let iter = tcx
+        .all_traits(LOCAL_CRATE)
+        .iter()
+        .copied()
+        .filter(|trait_| traits_in_scope.contains(trait_));
+    traits_implemented_impl(tcx, type_, iter)
+}
+
+fn traits_implemented_impl<'a>(
+    tcx: TyCtxt<'a>,
+    type_: DefId,
+    traits: impl Iterator<Item = DefId>,
+) -> &'a [DefId] {
+    let ty = tcx.type_of(type_);
+    let iter = traits.flat_map(|trait_| {
+        trace!("considering explicit impl for trait {:?}", trait_);
+        let mut saw_impl = false;
+        // Look at each trait implementation to see if it's an impl for `did`
+        tcx.for_each_relevant_impl(trait_, ty, |impl_| {
+            // TODO: find a way to short-circuit for_each_* so this doesn't take as long
+            if saw_impl {
+                return;
+            }
+
+            let trait_ref = tcx.impl_trait_ref(impl_).expect("this is not an inherent impl");
+            // Check if these are the same type.
+            let impl_type = trait_ref.self_ty();
+            debug!(
+                "comparing type {} with kind {:?} against type {:?}",
+                impl_type, impl_type.kind, type_
+            );
+            // Fast path: if this is a primitive simple `==` will work
+            saw_impl = impl_type == ty
+                || match impl_type.kind {
+                    // Check if these are the same def_id
+                    ty::Adt(def, _) => {
+                        debug!("adt def_id: {:?}", def.did);
+                        def.did == type_
+                    }
+                    ty::Foreign(def_id) => def_id == type_,
+                    _ => false,
+                };
+        });
+        if saw_impl { Some(trait_) } else { None }
+    });
+    tcx.arena.alloc_from_iter(iter)
```

Comment:
We probably want to return `ty::List<'tcx, DefId>` here so that we compare and hash by pointer.

Reply:
> A wrapper for slices with the additional invariant that the slice is interned and **no other slice with the same contents can exist in the same context**.

Are we sure no other slice with this contents can exist?

Reply:
Oh, I assumed you would be interning it -- I guess you're not doing that here. Given that these are probably only going to be used by rustdoc it probably doesn't matter too much.

Reply:
I actually think other parts of the compiler could benefit from this, e.g. anyone currently using `for_each_relevant_impl`. But for now it's enough to get it working for rustdoc.

Reply:
This is no longer relevant now that the caching is implemented in rustdoc and not a query.